pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -107,12 +107,22 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env model.Environmen
 			// this is for.
 			FilterChains: make([]plugin.FilterChain, len(l.FilterChains)),
 		}
+
+		var si *model.ServiceInstance
+		for _, w := range workloadInstances {
+			if w.Endpoint.Port == int(portNumber) {
+				si = w
+				break
+			}
+		}
+
```

Comment:
why do you need this? this loop is simply a merge of all user declared servers on the gateway pod(s). All pods of a gateway are required/expected to have set of all ports exposed the servers (80, 443, etc.). So, there is no case where one or more pods of a gateway have extra ports while other pods don't have. If anything, other code would break badly if this were to happen.
So I suggest removing this loop/check.
and in the input params below, pass workloadInstances[0]

Reply:
since we have the [for loop above to iterate all the ports](https://github.com/quanjielin/istio/blob/d3c01acb09793a2b1d1baf17dccf031e7eb6c251/pilot/pkg/networking/core/v1alpha3/gateway.go#L68), then build listener for each port; so the loop here is to find serviceInstance match to certain port, workloadInstances[0] cannot guarantee the match.

Reply:
All workload instances of the gateway are identical and have same ports. Unlike a normal sidecar enabled pod. We don’t guarantee correct behavior if some instances/pods of a gateway have different sets of ports from other instances. So the code above is inevitably picking [0] always. Might as well make it clear

Reply:
log shows that workloadInstances have different Endpoint.Port 

2018-06-27T00:36:08.908734Z	info	************buildGatewayListeners workloadInstance &{Endpoint:{Family:tcp Address:10.12.1.9 Port:80 ServicePort:0xc420b709c0 UID:} Service:0xc420b97280 Labels:istio=ingressgateway,pod-template-hash=3392915451 AvailabilityZone:us-central1/us-central1-a ServiceAccount:spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account} 

2018-06-27T00:36:08.908793Z	info	************buildGatewayListeners workloadInstance &{Endpoint:{Family:tcp Address:10.12.1.9 Port:31400 ServicePort:0xc420b70a20 UID:} Service:0xc420b97280 Labels:istio=ingressgateway,pod-template-hash=3392915451 AvailabilityZone:us-central1/us-central1-a ServiceAccount:spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account} 

2018-06-27T00:36:08.908829Z	info	************buildGatewayListeners workloadInstance &{Endpoint:{Family:tcp Address:10.12.1.9 Port:443 ServicePort:0xc420b709f0 UID:} Service:0xc420b97280 Labels:istio=ingressgateway,pod-template-hash=3392915451 AvailabilityZone:us-central1/us-central1-a ServiceAccount:spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account}

Reply:
Oh shoot. I totally forgot that we have one endpoint instance for each port in the gateway.

Reply:
Ignore all my nits.